examples/probability.dx

```diff
@@ -0,0 +1,297 @@
+
+' # Mini-PPL for Discrete variables in Dex
+
+' This document is an example for creating a small PPL in Dex. This is
+inspired by PPL's written for Haskell that utilize monadic
+represtations of the chain rule for representing discrete
+variables. Notably though in Dex we aim for a non-monadic style that
+primarily uses tables and autodifferentiation. 
+
+
+' ### General Helper functions for printing
+
+' These are some general functions to make printing easier
+ throughout the document. 
+
+def join (joiner: List a) (lists:n=>(List a)) : List a =
+    -- Join together lists with an intermediary joiner
+    concat $ for i.
+        case ordinal i == (size n - 1) of
+             True -> lists.i
+             False -> lists.i <> joiner
+
+instance Show (Fin a)
+  show = \ d. show $ ordinal d
+
+instance [Show a, Show b] Show (a & b)
+  show = \(a, b). "(" <> show a <> ", " <> show b <> ")"
+
+instance [Mul a, Mul b] Mul (a & b)
+     mul = \ (a, b) (a', b').  (a * a', b * b')
+     one = (one, one)
+
+
+' ## Distributions
+
+' A simple representation of a (possibly unnormalized) probability
+  table. As well as a alias for a conditional distribution.
+  Probs are defined as `data` to allow for sparse representations.
+
+data Probs a = AsProbsTable (a => Float)
+def CDist (a:Type) (c:Type): Type =  c => Probs a
+
+instance [Show a] Show (Probs a)
+  show = \ d.
+       case d of 
+         AsProbsTable d2 -> 
+                join "\n" $ for i.
+                   (show i <> " " <>  (concat $ for j: (Fin 50).
+                     select (((IToF (ordinal j)  + 1.0) / 50.0) < d2.i) "*" "" ) <> " " <> show d2.i)
+
+' Some useful distribution constructors
+
+(true, false) = (1@(Fin 2), 0@(Fin 2))
+Binary = Fin 2
+
+def bernoulli (p: Float) : Probs Binary = AsProbsTable [1.0 - p, p]
+
+def uniform : Probs m = AsProbsTable $ for i. 1.0 / (IToF (size m))
+
+def delta (x:m) : Probs m =
+    AsProbsTable $ yieldState zero $ \ aref.
+               aref!x := 1.0
+
+' ## Operations on probabilities
+
+
+' Construct a tuple from independent random variables -  $p(a, b) = p(a) p(b)$
+
+def (,,) (adist : Probs a) (bdist : Probs b) : Probs (a & b) =
+    (AsProbsTable a) = adist
+    (AsProbsTable b) = bdist 
+    AsProbsTable for (i,j). a.i * b.j
+
+' Apply the markov rule -  $p(a) = \sum_b p(a | b) p(b)$
+
+def (<|) (cond_dist : CDist a b) (prior : Probs b) : Probs a =
```

Comment:
I'm not sure if our parser supports it right now, but it would be kind of cool if you could write "a `ConditionedOn` b" and get a conditional distribution. Or you could use an operator for the type constructor and write something like `a || b`.